plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,464 @@
+	g_clear_object (&ctx->mbim_device);
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static gboolean
+fu_mbim_qdu_updater_reprobe_timeout (gpointer user_data)
+{
+	CloseContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+gboolean
+fu_mbim_qdu_updater_close (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	CloseContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = g_steal_pointer (&self->mbim_device),
+		.timeout_id = 0,
+	};
+
+	mbim_device_close (ctx.mbim_device, 5, NULL, fu_mbim_qdu_updater_mbim_device_close_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* we should always have both device cleared, and optionally error set */
+	g_warn_if_fail (ctx.mbim_device == NULL);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return FALSE;
+	}
+
+	g_warn_if_fail (ctx.timeout_id == 0);
+	ctx.timeout_id = g_timeout_add_seconds (15, fu_mbim_qdu_updater_reprobe_timeout, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* update detach right after this */
+	return TRUE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+  	GCancellable    *cancellable;
+	GError		*error;
+	guint		 timeout_id;
+	GBytes		*blob;
+	GArray		*digest;
+	GPtrArray	*chunks;
+	guint 		chunk_sent;
+	FuDevice 	*device;
+} WriteContext;
+
+static gboolean
+fu_mbim_qdu_updater_reboot_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(MbimMessage) request  = NULL;
+	g_autoptr(MbimMessage) response = NULL;
+	g_autoptr(GError)      error = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (response, &error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	fu_device_set_progress_full (FU_DEVICE (ctx->device), (gsize) ctx->chunk_sent, (gsize) ctx->chunks->len);
+	ctx->chunk_sent++;
+	if (ctx->chunk_sent < ctx->chunks->len) {
+		FuChunk *chk = g_ptr_array_index (ctx->chunks, ctx->chunk_sent);
+		request = mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), NULL);
+		mbim_device_command (ctx->mbim_device,
+				     request,
+				     20,
+				     ctx->cancellable,
+				     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+				     ctx);
+		mbim_message_unref (request);
```

Comment:
Maybe add a `return;` here and skip the else branch, reindenting its contents left?

Reply:
Maybe it's me and how I'm used to writing this kind of logic. Instead of an if/else at the end of the function like this:
```
    if (something) {
        do_something
    } else {
       g_debug (Done!);
       do_something_else
    }
} 
```

It could also be written like this:
```
    if (something) {
        do_something
        return;
    }

    g_debug (Done!);
    do_something_else
} 
```

I thought it made sense to write it like that because your else branch contained the final path that finishes the overall operation. Not a big deal really, I think it just could read better like that.

Reply:
OK, I understand what you mean, thanks for your patient explanation.